cogs/reminder.py

```diff
@@ -1,38 +1,145 @@
 
 
 class Reminder(commands.Cog):
+
     def __init__(self, bot: commands.Bot) -> None:
         self.bot = bot
+        self.active: Mapping[int, asyncio.Task] = {}
+    
+    async def handle_reminder(self, user: discord.User, reminder: MemberReminder):
+        """
+        Sleep until reminder's timestamp, then send the message to the user and delete the timer.
+        If cancelled, reminder deletion must be handled elsewhere.
+        """
+        message = reminder.message
+        terminus = reminder.timestamp
+        await long_sleep_until(terminus)
+        await user.send(f'You asked me to remind you: {message}')
+        await reminder.delete()
+    
+    def removal_callback(self, id: int):
+        """
+        Create callback for removing reminder's task from the internal mapping of active tasks.
+        To be used in asyncio.Task.add_done_callback when scheduling a timer.
+        """
+        def callback(task: asyncio.Task):
+            del self.active[id]
+        return callback
     
+    async def schedule_existing_reminders(self):
+        """
+        Schedule all timers existing in the database. To be used on bot start-up.
+        If a reminder's time has passed, it is scheduled to send with DEGENERACY_DELAY (to avoid rate-limiting).
+        """
+        reminders = await MemberReminder.all()
+        if not reminders:
+            logger.debug('No existing reminders found.')
+        
+        # on bot start-up, add a bit of delay between scheduled reminders (to avoid rate-limiting)
+        dormant = asyncio.create_task(
+            asyncio.sleep(DEGENERACY_DELAY.total_seconds())
+        )
+        async def schedule_once_completed(last: asyncio.Task, user: discord.User, reminder: MemberReminder):
+            await asyncio.wait_for(last, timeout=None)
+            await self.handle_reminder(user, reminder)
+
+        async with asyncio.TaskGroup() as tg:
+            for reminder in reminders:
+                if reminder.reminder_id in self.active:
+                    logger.debug('Reminder already active. (skipping)')
+                    continue
+                user = self.bot.get_user(reminder.user_id)
+                if not user:
+                    logger.warning(f'User {reminder.user_id} not found. (skipping)')
+                    continue
+
+                if reminder.timestamp <= timezone.now():
+                    dormant = tg.create_task(schedule_once_completed(dormant, user, reminder))
+                    logger.debug(f'Dormant timer running: {reminder}')
+                else:
+                    task = asyncio.create_task(self.handle_reminder(user, reminder))
+                    self.active[reminder.reminder_id] = task
+                    task.add_done_callback(self.removal_callback(reminder.reminder_id))
+                    logger.debug(f'Active timer running: {reminder}')
+
+                logger.debug(f'Active: {len(self.active)}')
```

Comment:
Might be worth changing this to `Active reminders` instead of just `Active` to easily read DEBUG logs.